src/frontend/Typechecker.ml

```diff
@@ -0,0 +1,1549 @@
+  match ue.emeta.type_ with UInt | UReal -> true | _ -> false
+
+let probability_distribution_name_variants id =
+  let name = id.name in
+  let open String in
+  List.map
+    ~f:(fun n -> {name= n; id_loc= id.id_loc})
+    ( if name = "multiply_log" || name = "binomial_coefficient_log" then [name]
+    else if is_suffix ~suffix:"_lpmf" name then
+      [name; drop_suffix name 5 ^ "_lpdf"; drop_suffix name 5 ^ "_log"]
+    else if is_suffix ~suffix:"_lpdf" name then
+      [name; drop_suffix name 5 ^ "_lpmf"; drop_suffix name 5 ^ "_log"]
+    else if is_suffix ~suffix:"_lcdf" name then
+      [name; drop_suffix name 5 ^ "_cdf_log"]
+    else if is_suffix ~suffix:"_lccdf" name then
+      [name; drop_suffix name 6 ^ "_ccdf_log"]
+    else if is_suffix ~suffix:"_cdf_log" name then
+      [name; drop_suffix name 8 ^ "_lcdf"]
+    else if is_suffix ~suffix:"_ccdf_log" name then
+      [name; drop_suffix name 9 ^ "_lccdf"]
+    else if is_suffix ~suffix:"_log" name then
+      [name; drop_suffix name 4 ^ "_lpmf"; drop_suffix name 4 ^ "_lpdf"]
+    else [name] )
+
+(* -- General checks ---------------------------------------------- *)
+let reserved_keywords =
+  [ "true"; "false"; "repeat"; "until"; "then"; "var"; "fvar"; "STAN_MAJOR"
+  ; "STAN_MINOR"; "STAN_PATCH"; "STAN_MATH_MAJOR"; "STAN_MATH_MINOR"
+  ; "STAN_MATH_PATCH"; "alignas"; "alignof"; "and"; "and_eq"; "asm"; "auto"
+  ; "bitand"; "bitor"; "bool"; "break"; "case"; "catch"; "char"; "char16_t"
+  ; "char32_t"; "class"; "compl"; "const"; "constexpr"; "const_cast"
+  ; "continue"; "decltype"; "default"; "delete"; "do"; "double"; "dynamic_cast"
+  ; "else"; "enum"; "explicit"; "export"; "extern"; "false"; "float"; "for"
+  ; "friend"; "goto"; "if"; "inline"; "int"; "long"; "mutable"; "namespace"
+  ; "new"; "noexcept"; "not"; "not_eq"; "nullptr"; "operator"; "or"; "or_eq"
+  ; "private"; "protected"; "public"; "register"; "reinterpret_cast"; "return"
+  ; "short"; "signed"; "sizeof"; "static"; "static_assert"; "static_cast"
+  ; "struct"; "switch"; "template"; "this"; "thread_local"; "throw"; "true"
+  ; "try"; "typedef"; "typeid"; "typename"; "union"; "unsigned"; "using"
+  ; "virtual"; "void"; "volatile"; "wchar_t"; "while"; "xor"; "xor_eq"
+  ; "functions"; "data"; "parameters"; "model"; "return"; "if"; "else"; "while"
+  ; "for"; "in"; "break"; "continue"; "void"; "int"; "real"; "complex"
+  ; "vector"; "row_vector"; "matrix"; "ordered"; "positive_ordered"; "simplex"
+  ; "unit_vector"; "cholesky_factor_corr"; "cholesky_factor_cov"; "corr_matrix"
+  ; "cov_matrix"; "print"; "reject"; "target"; "get_lp"; "profile" ]
+
+let verify_identifier id : unit =
+  if id.name = !model_name then
+    Semantic_error.ident_is_model_name id.id_loc id.name |> error
+  else if
+    String.is_suffix id.name ~suffix:"__"
+    || List.mem reserved_keywords id.name ~equal:String.equal
+  then Semantic_error.ident_is_keyword id.id_loc id.name |> error
+  else ()
+
+let verify_name_fresh_var tenv id =
```

Comment:
We currently have a few of those, though a lot of them had gone stale which made this refactor difficult. In general I preserved any comments which still seemed valid. I want to avoid commenting every function with what might be an obvious statement, so if there are any specifically (like this one) that weren't clear I'd love to comment just those ones.

These two functions check that a name used in a declaration is previously on unused. They're different because we allow variable names to shadow stanlib, but not UDFs. I'll comment the distinction